pkg/services/oauthserver/oasimpl/service.go

```diff
@@ -0,0 +1,504 @@
+}
+
+// saveServiceAccount creates a service account if the service account ID is NoServiceAccountID, otherwise it updates the service account's permissions
+func (s *OAuth2ServiceImpl) saveServiceAccount(ctx context.Context, extSvcName string, saID int64, permissions []ac.Permission) (int64, error) {
+	if saID == oauthserver.NoServiceAccountID {
+		// Create a service account
+		s.logger.Debug("Create service account", "external service name", extSvcName)
+		return s.createServiceAccount(ctx, extSvcName, permissions)
+	}
+
+	// check if the service account exists
+	s.logger.Debug("Update service account", "external service name", extSvcName)
+	sa, err := s.saService.RetrieveServiceAccount(ctx, oauthserver.TmpOrgID, saID)
+	if err != nil {
+		s.logger.Error("Error retrieving service account", "external service name", extSvcName, "error", err)
+		return oauthserver.NoServiceAccountID, err
+	}
+
+	// update the service account's permissions
+	if len(permissions) > 0 {
+		s.logger.Debug("Update role permissions", "external service name", extSvcName, "saID", saID)
+		if err := s.acService.SaveExternalServiceRole(ctx, ac.SaveExternalServiceRoleCommand{
+			OrgID:             ac.GlobalOrgID,
+			Global:            true,
+			ExternalServiceID: extSvcName,
+			ServiceAccountID:  sa.Id,
+			Permissions:       permissions,
+		}); err != nil {
+			return oauthserver.NoServiceAccountID, err
+		}
+		return saID, nil
+	}
+
+	// remove the service account
+	errDelete := s.deleteServiceAccount(ctx, extSvcName, sa.Id)
+	return oauthserver.NoServiceAccountID, errDelete
+}
+
+// deleteServiceAccount deletes a service account by ID and removes its associated role
+func (s *OAuth2ServiceImpl) deleteServiceAccount(ctx context.Context, extSvcName string, saID int64) error {
+	s.logger.Debug("Delete service account", "external service name", extSvcName, "saID", saID)
+	if err := s.saService.DeleteServiceAccount(ctx, oauthserver.TmpOrgID, saID); err != nil {
+		return err
+	}
+	return s.acService.DeleteExternalServiceRole(ctx, extSvcName)
+}
+
+// createServiceAccount creates a service account with the given permissions
+// and returns the ID of the service account
+// When no permission is given, the account isn't created and NoServiceAccountID is returned
+// This first design does not use a single transaction for the whole service account creation process => database consistency is not guaranteed.
+// Consider changing this in the future.
+func (s *OAuth2ServiceImpl) createServiceAccount(ctx context.Context, extSvcName string, permissions []ac.Permission) (int64, error) {
+	if len(permissions) == 0 {
+		// No permissions, no service account
+		s.logger.Debug("No permissions, no service account", "external service name", extSvcName)
+		return oauthserver.NoServiceAccountID, nil
+	}
+
+	newRole := func(r roletype.RoleType) *roletype.RoleType {
+		return &r
+	}
+	newBool := func(b bool) *bool {
+		return &b
+	}
+
+	slug := slugify.Slugify(extSvcName)
```

Comment:
I wonder if instead of the slug (based on the name <=> external input) we should use the client ID (random and generated by us), that would probably be more secure and prevent service account take over :thinking:

Reply:
Agree! Using clientID would be better

Reply:
Ok, there are a couple of issues with this given we re-gen the client ID upon save (done on every startup for a plugin using this). This would currently mean changing the service account altogether. So I'd suggest tracking this in another issue & PR

Reply:
Tracked in an issue.